### PR TITLE
feature: ip adapter masking

### PR DESCRIPTION
**Pull request by isamu-isozaki**
This is a WIP pr. I checked https://github.com/cubiq/ComfyUI_IPAdapter_plus/tree/main
and my understanding of how the masking is done is as follows
1. Each image and mask is sent through the IP adapter. When the IP adapter finishes calculating the attention, they multiply by the resized mask. This is not done at the qk phase but after the qkv computation as a whole is done.
2. Just sum all those new image prompts to the base text cross-attention output.

I'm pretty sure the above is the workflow but let me know if anyone thinks it's mistaken!
The main tasks left for this to be ready for review is

- [x] Using refiner's Context, or another way, figure out how to pass a mask as an argument to IPScaledDotProductAttention(init concept kind of done.)
- [x] Figure out how to make the same projection layers for k and v operate on more than 1 image prompt(proof of concept idea done)
- [x] Test on sd1.5

### CODE REVIEW

File: src/refiners/foundationals/latent_diffusion/stable_diffusion_1/unet.py [hunk 1]
```diff
@@ -280,10 +280,13 @@ def init_context(self) -> Contexts:
...
             "sampling": {"shapes": []},
+            "ip_mask": {"mask": None},
         }
 
     def set_clip_text_embedding(self, clip_text_embedding: Tensor) -> None:
         self.set_context("cross_attention_block", {"clip_text_embedding": clip_text_embedding})
 
     def set_timestep(self, timestep: Tensor) -> None:
         self.set_context("diffusion", {"timestep": timestep})
+    def set_ip_mask(self, mask: Tensor) -> None:
```
Comment (limiteinductive) [inline src/refiners/foundationals/latent_diffusion/stable_diffusion_1/unet.py:291]: I think this method should belong to IPAdapter. One of the core design patterns of Refiners is to try to modify the previous code as little as possible. In this case, unet.py should know very little about IPAdapter.

Comment (isamu-isozaki) [reply src/refiners/foundationals/latent_diffusion/stable_diffusion_1/unet.py:291]: Thanks for the check! Ah let me learn more about the context API haha

Comment (limiteinductive) [reply src/refiners/foundationals/latent_diffusion/stable_diffusion_1/unet.py:291]: We should probably write a notebook to clarify how the Context API is working 😅

Comment (isamu-isozaki) [reply src/refiners/foundationals/latent_diffusion/stable_diffusion_1/unet.py]: Ok. I think I did this

**Comment by isamu-isozaki** [comment]
For now I'll try cleaning code with type stuff and see if I missed anything

**Comment by isamu-isozaki** [comment]
Ok! Might have finally done it. 
double_mask

**Comment by isamu-isozaki** [comment]
Ok! Here's my current best results:
more_minimal_test-plus

I noticed that the IP adapter plus is needed to get descent results with masks but also, even on the default branch, at least with fp16, I have a bit of a fuzzi-ness like below
original_-plus

The code for running the double mask is below
```
import torch
import numpy as np
from PIL import Image
import os
from refiners.foundationals.latent_diffusion.stable_diffusion_1 import StableDiffusion_1
from refiners.foundationals.latent_diffusion.freeu import SDFreeUAdapter
from refiners.foundationals.latent_diffusion.schedulers.dpm_solver import DPMSolver
from refiners.fluxion.utils import manual_seed, image_to_tensor, load_from_safetensors
from refiners.foundationals.latent_diffusion import SD1IPAdapter
version = "-plus"
postfix = ""
torch_dtype=torch.float16
sd1 = StableDiffusion_1(device="cuda", dtype=torch_dtype)
model_path = r"..\refiners\tests\weights"
sd1.lda.load_from_safetensors(os.path.join(model_path, "lda.safetensors"))
sd1.unet.load_from_safetensors(os.path.join(model_path, "unet.safetensors"))
sd1.clip_text_encoder.load_from_safetensors(os.path.join(model_path, "CLIPTextEncoderL.safetensors"))
ip_adapter = SD1IPAdapter(target=sd1.unet, weights=load_from_safetensors(os.path.join(model_path, f"ip-adapter{version}_sd15-multi.safetensors")), num_image_prompts=2, fine_grained=(version=="-plus"))
ip_adapter.clip_image_encoder.load_from_safetensors(os.path.join(model_path, "CLIPImageEncoderH-multi.safetensors"))
ip_adapter.inject()

image_files=["knight.jpg", "painting.jpg"]
image_size=224
latent_size = (512, 768)
mask_width=latent_size[-1]//2
image_prompts = [Image.open(image_file).resize((image_size, image_size)) for image_file in image_files]
masks = [np.zeros((2, *latent_size)) for _ in range(2)]
masks[0][:, :, :mask_width] = 1
masks[1][:, :, mask_width:] = 1
torch_masks = [torch.tensor(mask).to("cuda", dtype=torch_dtype) for mask in masks]

def test_prompt(prompt, output_file, image_prompts, masks, output_size=(512, 512), ip_adapter_scale=0.7):
    
    # Tune parameters
    seed = 9752
    first_step = 1
    ip_adapter.set_scale(ip_adapter_scale)
    sd1.set_num_inference_steps(50)
    with torch.no_grad():
        # Note: default text prompts for IP-Adapter
        clip_text_embedding = sd1.compute_clip_text_embedding(
            text=f"high quality, highly detailed {prompt}", negative_text="blurry, monochrome, lowres, bad anatomy, worst quality, low quality"
        )
        clip_image_embeddings = [ip_adapter.compute_clip_image_embedding(ip_adapter.preprocess_image(image_prompt)) for image_prompt in image_prompts]
        for clip_image_embedding in clip_image_embeddings:
            negative_text_embedding, conditional_text_embedding = clip_text_embedding.chunk(2)
            negative_image_embedding, conditional_image_embedding = clip_image_embedding.chunk(2)
            clip_text_embedding = torch.cat(
                (
                    torch.cat([negative_text_embedding, negative_image_embedding], dim=1),
                    torch.cat([conditional_text_embedding, conditional_image_embedding], dim=1),
                )
            )
        def context_callback():
            ip_adapter.set_mask(tuple(masks))
        manual_seed(seed=seed)
        x = sd1.init_latents(size=output_size, first_step=first_step).to(
            device=sd1.device, dtype=sd1.dtype
        )
        for step in sd1.steps[first_step:]:
            x = sd1(
                x,
                step=step,
                clip_text_embedding=clip_text_embedding,
                context_callback=context_callback
            )
        predicted_image = sd1.lda.decode_latents(x=x)
    
    predicted_image.save(output_file)
    print(f"done: see {output_file}")
test_prompt("two girl friends laughing", f"ip_adapter/more_minimal_test{version}{postfix}.jpg", image_prompts, torch_masks, output_size=latent_size)
```

**Comment by isamu-isozaki** [comment]
Oh, I was completely wrong. I just needed lower cfg haha
more_minimal_test_lower_cond-plus

**Comment by isamu-isozaki** [comment]
Ok! Found the correct weights [here](https://gist.github.com/deltheil/0e55b1b625e0b292239e9e42a69d9d4e) by @deltheil . Seems like the default sd doesn't work with this IP adapter as far as I've tested. But with the new weights:
more_minimal_test_lower_cond-plus
more_minimal_test-plus

I found the original issue author [here](https://www.youtube.com/watch?v=vqG1VXKteQg&ab_channel=LatentVision) and he seems to use DreamShaper

**Comment by deltheil** [comment]
> Found the correct weights [here](https://gist.github.com/deltheil/0e55b1b625e0b292239e9e42a69d9d4e) by @deltheil . Seems like the default sd doesn't work with this IP adapter as far as I've tested

Cool! FYI, the goal was to reproduce https://github.com/tencent-ailab/IP-Adapter/blob/main/ip_adapter-plus_demo.ipynb which uses `SG161222/Realistic_Vision_V4.0_noVAE`, hence the gist.

Have you also tried with `sd15/dreamshaper_8.safetensors`?

**Comment by isamu-isozaki** [comment]
@deltheil nope! Not yet. Let me try now

**Comment by isamu-isozaki** [comment]
@deltheil It seems like better compositio
dream_more_minimal_test_lower_cond-plus
n

**Comment by isamu-isozaki** [comment]
Just for reference, with advice from @limiteinductive the current test function is this
```
import torch
import numpy as np
from PIL import Image
import os
from refiners.foundationals.latent_diffusion.stable_diffusion_1 import StableDiffusion_1
from refiners.foundationals.latent_diffusion.freeu import SDFreeUAdapter
from refiners.foundationals.latent_diffusion.schedulers.dpm_solver import DPMSolver
from refiners.fluxion.utils import manual_seed, image_to_tensor, load_from_safetensors
from refiners.foundationals.latent_diffusion import SD1IPAdapter
version = "-plus"
postfix = ""
torch_dtype=torch.float16
sd1 = StableDiffusion_1(device="cuda", dtype=torch_dtype)
model_path = r"..\refiners\tests\weights"
sd1.lda.load_from_safetensors(os.path.join(model_path, "lda.safetensors"))
sd1.unet.load_from_safetensors(os.path.join(model_path, "unet.safetensors"))
sd1.clip_text_encoder.load_from_safetensors(os.path.join(model_path, "CLIPTextEncoderL.safetensors"))
ip_adapter = SD1IPAdapter(target=sd1.unet, weights=load_from_safetensors(os.path.join(model_path, f"ip-adapter{version}_sd15-multi.safetensors")), num_image_prompts=2, fine_grained=(version=="-plus"))
ip_adapter.clip_image_encoder.load_from_safetensors(os.path.join(model_path, "CLIPImageEncoderH-multi.safetensors"))
ip_adapter.inject()

image_files=["knight.jpg", "painting.jpg"]
image_size=224
latent_size = (512, 768)
mask_width=latent_size[-1]//2
image_prompts = [Image.open(image_file).resize((image_size, image_size)) for image_file in image_files]
masks = [np.zeros((2, *latent_size)) for _ in range(2)]
masks[0][:, :, :mask_width] = 1
masks[1][:, :, mask_width:] = 1
torch_masks = [torch.tensor(mask).to("cuda", dtype=torch_dtype) for mask in masks]

def test_prompt(prompt, output_file, image_prompts, masks, output_size=(512, 512), ip_adapter_scale=0.7):
    
    # Tune parameters
    seed = 9752
    first_step = 1
    ip_adapter.set_scale(ip_adapter_scale)
    sd1.set_num_inference_steps(50)
    with torch.no_grad():
        # Note: default text prompts for IP-Adapter
        clip_text_embedding = sd1.compute_clip_text_embedding(
            text=f"high quality, highly detailed {prompt}", negative_text="blurry, monochrome, lowres, bad anatomy, worst quality, low quality"
        )
        clip_image_embeddings = [ip_adapter.compute_clip_image_embedding(ip_adapter.preprocess_image(image_prompt)) for image_prompt in image_prompts]
        for clip_image_embedding in clip_image_embeddings:
            negative_text_embedding, conditional_text_embedding = clip_text_embedding.chunk(2)
            negative_image_embedding, conditional_image_embedding = clip_image_embedding.chunk(2)
            clip_text_embedding = torch.cat(
                (
                    torch.cat([negative_text_embedding, negative_image_embedding], dim=1),
                    torch.cat([conditional_text_embedding, conditional_image_embedding], dim=1),
                )
            )
        ip_adapter.set_mask(tuple(masks))
        manual_seed(seed=seed)
        x = sd1.init_latents(size=output_size, first_step=first_step).to(
            device=sd1.device, dtype=sd1.dtype
        )
        for step in sd1.steps[first_step:]:
            x = sd1(
                x,
                step=step,
                clip_text_embedding=clip_text_embedding
            )
        predicted_image = sd1.lda.decode_latents(x=x)
    
    predicted_image.save(output_file)
    print(f"done: see {output_file}")
test_prompt("two girl friends laughing", f"ip_adapter/more_minimal_test{version}{postfix}.jpg", image_prompts, torch_masks, output_size=latent_size)
```
where we can just do ip_adapter.set_mask(tuple(masks)) instead.

**Comment by limiteinductive** [comment]
> Just for reference, with advice from @limiteinductive the current test function is this
> 
> ```
> import torch
> import numpy as np
> from PIL import Image
> import os
> from refiners.foundationals.latent_diffusion.stable_diffusion_1 import StableDiffusion_1
> from refiners.foundationals.latent_diffusion.freeu import SDFreeUAdapter
> from refiners.foundationals.latent_diffusion.schedulers.dpm_solver import DPMSolver
> from refiners.fluxion.utils import manual_seed, image_to_tensor, load_from_safetensors
> from refiners.foundationals.latent_diffusion import SD1IPAdapter
> version = "-plus"
> postfix = ""
> torch_dtype=torch.float16
> sd1 = StableDiffusion_1(device="cuda", dtype=torch_dtype)
> model_path = r"..\refiners\tests\weights"
> sd1.lda.load_from_safetensors(os.path.join(model_path, "lda.safetensors"))
> sd1.unet.load_from_safetensors(os.path.join(model_path, "unet.safetensors"))
> sd1.clip_text_encoder.load_from_safetensors(os.path.join(model_path, "CLIPTextEncoderL.safetensors"))
> ip_adapter = SD1IPAdapter(target=sd1.unet, weights=load_from_safetensors(os.path.join(model_path, f"ip-adapter{version}_sd15-multi.safetensors")), num_image_prompts=2, fine_grained=(version=="-plus"))
> ip_adapter.clip_image_encoder.load_from_safetensors(os.path.join(model_path, "CLIPImageEncoderH-multi.safetensors"))
> ip_adapter.inject()
> 
> image_files=["knight.jpg", "painting.jpg"]
> image_size=224
> latent_size = (512, 768)
> mask_width=latent_size[-1]//2
> image_prompts = [Image.open(image_file).resize((image_size, image_size)) for image_file in image_files]
> masks = [np.zeros((2, *latent_size)) for _ in range(2)]
> masks[0][:, :, :mask_width] = 1
> masks[1][:, :, mask_width:] = 1
> torch_masks = [torch.tensor(mask).to("cuda", dtype=torch_dtype) for mask in masks]
> 
> def test_prompt(prompt, output_file, image_prompts, masks, output_size=(512, 512), ip_adapter_scale=0.7):
>     
>     # Tune parameters
>     seed = 9752
>     first_step = 1
>     ip_adapter.set_scale(ip_adapter_scale)
>     sd1.set_num_inference_steps(50)
>     with torch.no_grad():
>         # Note: default text prompts for IP-Adapter
>         clip_text_embedding = sd1.compute_clip_text_embedding(
>             text=f"high quality, highly detailed {prompt}", negative_text="blurry, monochrome, lowres, bad anatomy, worst quality, low quality"
>         )
>         clip_image_embeddings = [ip_adapter.compute_clip_image_embedding(ip_adapter.preprocess_image(image_prompt)) for image_prompt in image_prompts]
>         for clip_image_embedding in clip_image_embeddings:
>             negative_text_embedding, conditional_text_embedding = clip_text_embedding.chunk(2)
>             negative_image_embedding, conditional_image_embedding = clip_image_embedding.chunk(2)
>             clip_text_embedding = torch.cat(
>                 (
>                     torch.cat([negative_text_embedding, negative_image_embedding], dim=1),
>                     torch.cat([conditional_text_embedding, conditional_image_embedding], dim=1),
>                 )
>             )
>         ip_adapter.set_mask(tuple(masks))
>         manual_seed(seed=seed)
>         x = sd1.init_latents(size=output_size, first_step=first_step).to(
>             device=sd1.device, dtype=sd1.dtype
>         )
>         for step in sd1.steps[first_step:]:
>             x = sd1(
>                 x,
>                 step=step,
>                 clip_text_embedding=clip_text_embedding
>             )
>         predicted_image = sd1.lda.decode_latents(x=x)
>     
>     predicted_image.save(output_file)
>     print(f"done: see {output_file}")
> test_prompt("two girl friends laughing", f"ip_adapter/more_minimal_test{version}{postfix}.jpg", image_prompts, torch_masks, output_size=latent_size)
> ```
> 
> where we can just do ip_adapter.set_mask(tuple(masks)) instead.

Hey, I'm taking a look at it. Could you rebase it so it's up to date with main?

**Comment by isamu-isozaki** [comment]
@limiteinductive thanks! On my end I'll start doing the pyright stuff. For style I think I ran all the other stuff

**Comment by isamu-isozaki** [comment]
Slightly changed the IPScaledDotProduct and ParallelLinear logic to work with pyrange but let me know what you think! Currently finished all the poetry runs including pyrange

**Comment by limiteinductive** [comment]
I tried your code with RealisticVision4 and I get this images that seems to make two knights?
image

**Comment by limiteinductive** [comment]
> I tried your code with RealisticVision4 and got this image that seems to make two knights. image

the second image gets ignored, if I put a cat, I get the same result.
image

**Comment by isamu-isozaki** [comment]
@limiteinductive I checked the original implementation and they used feather masks which seemed to fix this issue. Also, changing the prompt for the cat is needed too I think. For example, I tried
cat
knight

the above images and I got 
dream_more_minimal_test_lower_cond_cat_girl_feather-plus

for dreamshaper and
realistic_more_minimal_test_lower_cond_cat_girl_feather-plus
for the realistic one!
Without both the prompt change and the feather masks, it becomes 2 knights
The updated code is
```
import torch
import numpy as np
from PIL import Image
import os
from refiners.foundationals.latent_diffusion.stable_diffusion_1 import StableDiffusion_1
from refiners.foundationals.latent_diffusion.freeu import SDFreeUAdapter
from refiners.foundationals.latent_diffusion.schedulers.dpm_solver import DPMSolver
from refiners.fluxion.utils import manual_seed, image_to_tensor, load_from_safetensors
from refiners.foundationals.latent_diffusion import SD1IPAdapter
version = "-plus"
postfix = ""
torch_dtype=torch.float16
sd1 = StableDiffusion_1(device="cuda", dtype=torch_dtype)
model_path = r"..\refiners\tests\weights"
sd1.lda.load_from_safetensors(os.path.join(model_path, "lda.safetensors"))
sd1.unet.load_from_safetensors(os.path.join(model_path, "unet.safetensors"))
sd1.clip_text_encoder.load_from_safetensors(os.path.join(model_path, "CLIPTextEncoderL.safetensors"))
ip_adapter = SD1IPAdapter(target=sd1.unet, weights=load_from_safetensors(os.path.join(model_path, f"ip-adapter{version}_sd15-multi.safetensors")), num_image_prompts=2, fine_grained=(version=="-plus"))
ip_adapter.clip_image_encoder.load_from_safetensors(os.path.join(model_path, "CLIPImageEncoderH-multi.safetensors"))
ip_adapter.inject()

image_files=["knight.jpg", "painting.jpg"]
image_size=224
latent_size = (512, 768)
mask_width=latent_size[-1]//2
image_prompts = [Image.open(image_file).resize((image_size, image_size)) for image_file in image_files]
masks = [np.zeros((2, *latent_size)) for _ in range(2)]
mask_width=latent_size[-1]//4
half_width = latent_size[-1]//2
masks = [np.zeros((2, *latent_size)) for _ in range(2)]
masks[0][:, :, :half_width] = 1
masks[1][:, :, -half_width:] = 1

for i in range(mask_width, half_width+1):
    # at mask_width it's 1 and at half width it's 0
    # and linear in between
    current = 1-(i-mask_width)/(half_width-mask_width)
    masks[0][:, :, i] = current
    masks[1][:, :, -i] = current
torch_masks = [torch.tensor(mask).to("cuda", dtype=torch_dtype) for mask in masks]
mask_width=latent_size[-1]//2

def test_prompt(prompt, output_file, image_prompts, masks, output_size=(512, 512), ip_adapter_scale=0.7):
    
    # Tune parameters
    seed = 9752
    first_step = 1
    ip_adapter.set_scale(ip_adapter_scale)
    sd1.set_num_inference_steps(50)
    with torch.no_grad():
        # Note: default text prompts for IP-Adapter
        clip_text_embedding = sd1.compute_clip_text_embedding(
            text=f"high quality, highly detailed {prompt}", negative_text="blurry, monochrome, lowres, bad anatomy, worst quality, low quality"
        )
        clip_image_embeddings = [ip_adapter.compute_clip_image_embedding(ip_adapter.preprocess_image(image_prompt)) for image_prompt in image_prompts]
        for clip_image_embedding in clip_image_embeddings:
            negative_text_embedding, conditional_text_embedding = clip_text_embedding.chunk(2)
            negative_image_embedding, conditional_image_embedding = clip_image_embedding.chunk(2)
            clip_text_embedding = torch.cat(
                (
                    torch.cat([negative_text_embedding, negative_image_embedding], dim=1),
                    torch.cat([conditional_text_embedding, conditional_image_embedding], dim=1),
                )
            )
        ip_adapter.set_mask(tuple(masks))
        manual_seed(seed=seed)
        x = sd1.init_latents(size=output_size, first_step=first_step).to(
            device=sd1.device, dtype=sd1.dtype
        )
        for step in sd1.steps[first_step:]:
            x = sd1(
                x,
                step=step,
                clip_text_embedding=clip_text_embedding
            )
        predicted_image = sd1.lda.decode_latents(x=x)
    
    predicted_image.save(output_file)
    print(f"done: see {output_file}")
test_prompt("a knight and a cat", f"ip_adapter/more_minimal_test{version}{postfix}.jpg", image_prompts, torch_masks, output_size=latent_size)
```
Let me know if this fixes your issue!

**Comment by isamu-isozaki** [comment]
Ok. I thought I adapted to the new changes but now I'm getting images like below
dream_more_minimal_test_lower_cond_cat_girl_feather2-plus

**Comment by isamu-isozaki** [comment]
Ok. By messing with the masks I get:
dream_default_cat_girl-plus

**Comment by isamu-isozaki** [comment]
since diffusers have an implementation now I think I'll see if I can use that vs this to see if the implementation is correct

**Comment by isamu-isozaki** [comment]
I ran a test in diffusers and currently updating my code to newest changes so I can test on this branch too. The test in diffusers was:
```python
from diffusers import AutoPipelineForText2Image, DDIMScheduler
import torch
from diffusers.utils import load_image
from transformers import CLIPVisionModelWithProjection
from diffusers.image_processor import IPAdapterMaskProcessor
from refiners.fluxion.utils import manual_seed
noise_scheduler = DDIMScheduler(
    num_train_timesteps=1000,
    beta_start=0.00085,
    beta_end=0.012,
    beta_schedule="scaled_linear",
    clip_sample=False,
    set_alpha_to_one=False,
    steps_offset=1
)

pipeline = AutoPipelineForText2Image.from_pretrained(
    "SG161222/Realistic_Vision_V4.0_noVAE",
    torch_dtype=torch.float16,
    scheduler=noise_scheduler,
    feature_extractor=None,
    safety_checker=None
).to("cuda")
pipeline.load_ip_adapter("h94/IP-Adapter", subfolder="models", weight_name=["ip-adapter-plus_sd15.bin"]*2)

pipeline.scheduler = DDIMScheduler.from_config(pipeline.scheduler.config)
image1 = load_image("https://huggingface.co/datasets/YiYiXu/testing-images/resolve/main/ip_mask_girl1.png")
image2 = load_image("https://huggingface.co/datasets/YiYiXu/testing-images/resolve/main/ip_mask_girl2.png")
mask1 = load_image("https://huggingface.co/datasets/YiYiXu/testing-images/resolve/main/ip_mask_mask1.png")
mask2 = load_image("https://huggingface.co/datasets/YiYiXu/testing-images/resolve/main/ip_mask_mask2.png")

processor = IPAdapterMaskProcessor()
masks = processor.preprocess([mask1, mask2])
pipeline.set_ip_adapter_scale([0.5, 0.5])

ip_images =[[image1], [image2]]

manual_seed(9752)
num_images=1

images = pipeline(
    prompt="2 girls",
    ip_adapter_image=ip_images,
    negative_prompt="monochrome, lowres, bad anatomy, worst quality, low quality", 
    num_inference_steps=20, num_images_per_prompt=num_images, 
    cross_attention_kwargs={"ip_adapter_masks": masks}
).images
```
which from
image

and masks
image

I get

image

**Comment by isamu-isozaki** [comment]
@limiteinductive got same result! I think I will clean this up a bit and then I should be done
vs_diffusers_-plus

**Comment by isamu-isozaki** [comment]
The code for refiners is this:
```python
import torch

from PIL import Image
import os
from refiners.foundationals.latent_diffusion.stable_diffusion_1 import StableDiffusion_1
from refiners.fluxion.utils import manual_seed, image_to_tensor, load_from_safetensors
from refiners.foundationals.latent_diffusion import SD1IPAdapter
from refiners.foundationals.latent_diffusion.solvers.ddim import DDIM, SolverParams
from refiners.foundationals.latent_diffusion.solvers import TimestepSpacing, NoiseSchedule, ModelPredictionType

from diffusers.image_processor import IPAdapterMaskProcessor
from diffusers.utils import load_image
import numpy as np
params = SolverParams(
    num_train_timesteps=1000,
    timesteps_spacing=TimestepSpacing.LINSPACE_ROUNDED,
    timesteps_offset=0,
    initial_diffusion_rate=8.5e-4,
    final_diffusion_rate=1.2e-2,
    noise_schedule=NoiseSchedule.QUADRATIC,
    model_prediction_type=ModelPredictionType.NOISE,
)
version = "-plus"
postfix = ""
#prefix ="dream_"
prefix="realistic_"
torch_dtype=torch.float16
sd1 = StableDiffusion_1(device="cuda", dtype=torch_dtype)
model_path = r"..\refiners\tests\weights"
sd1.lda.load_from_safetensors(os.path.join(model_path, "lda_ft_mse.safetensors"))
sd1.unet.load_from_safetensors(os.path.join(model_path, f"{prefix}unet.safetensors"))
sd1.clip_text_encoder.load_from_safetensors(os.path.join(model_path, f"{prefix}clip_text.safetensors"))
sd1.solver = DDIM(num_inference_steps=20, params=params).to(device="cuda", dtype=torch_dtype)
ip_adapter = SD1IPAdapter(target=sd1.unet, weights=load_from_safetensors(os.path.join(model_path, f"ip-adapter{version}_sd15.safetensors")), num_image_prompts=2, fine_grained=(version=="-plus"))
ip_adapter.clip_image_encoder.load_from_safetensors(os.path.join(model_path, "clip_image.safetensors"))
ip_adapter.inject()

image1 = load_image("https://huggingface.co/datasets/YiYiXu/testing-images/resolve/main/ip_mask_girl1.png")
image2 = load_image("https://huggingface.co/datasets/YiYiXu/testing-images/resolve/main/ip_mask_girl2.png")
mask1 = load_image("https://huggingface.co/datasets/YiYiXu/testing-images/resolve/main/ip_mask_mask1.png")
mask2 = load_image("https://huggingface.co/datasets/YiYiXu/testing-images/resolve/main/ip_mask_mask2.png")

processor = IPAdapterMaskProcessor()
masks = processor.preprocess([mask1, mask2])
masks = masks.to("cuda", torch.float16)
masks = masks.repeat(1, 2, 1, 1)
masks = masks.reshape(4, masks.shape[2], masks.shape[3])
masks = masks.chunk(2)
image_prompts = [image1, image2]
latent_size = (512, 512)
def test_prompt(prompt, output_file, image_prompts, masks, output_size=(512, 512), ip_adapter_scale=0.7, condition_scale=7):
    
    # Tune parameters
    seed = 9752
    first_step = 1
    ip_adapter.scale = ip_adapter_scale
    sd1.set_inference_steps(20)
    with torch.no_grad():
        # Note: default text prompts for IP-Adapter
        clip_text_embedding = sd1.compute_clip_text_embedding(
            text=f"{prompt}", negative_text="blurry, monochrome, lowres, bad anatomy, worst quality, low quality"
        )
        clip_image_embeddings = tuple([ip_adapter.compute_clip_image_embedding(ip_adapter.preprocess_image(image_prompt)) for image_prompt in image_prompts])
        ip_adapter.set_mask(tuple(masks))
        ip_adapter.set_clip_image_embedding_from_tuple(clip_image_embeddings)
        manual_seed(seed=seed)
        x = sd1.init_latents(size=output_size).to(
            device=sd1.device, dtype=sd1.dtype
        )
        for step in sd1.steps[first_step:]:
            x = sd1(
                x,
                step=step,
                clip_text_embedding=clip_text_embedding,
                condition_scale=condition_scale,
            )
        predicted_image = sd1.lda.decode_latents(x=x)
    
    predicted_image.save(output_file)
    print(f"done: see {output_file}")
    return predicted_image
test_prompt("2 girls", f"ip_adapter/vs_diffusers_{version}{postfix}.jpg", image_prompts, masks, output_size=latent_size, ip_adapter_scale=0.5)
```

**Comment by isamu-isozaki** [comment]
Sorry, I got sidetracked by finals. I think the main issue with this PR is that how the mask is passed in is counter-intuitive so let me fix that

**Pull request closed without merging**